### PR TITLE
Add support for AD calculations to FEValues functions 

### PR DESCRIPTION
**Pull request by jppelteret**
When using AD to automatically compute the linearisation of a residual, one cannot _a priori_ assume no sensitivity with respect to zero-valued degrees-of-freedom. The first commit in this PR ensures that this calculation shortcut is not taken for AD numbers.

Furthermore, with the current functionality my application of these two features in the FEValues function is now overly defensive. So the second commit effectively reverts some of the changes made in #4951, as #4893 makes some use of `ProductType` unnecessary and #5251 ensures compatible conversions between numbers of all CV qualifications.

### CODE REVIEW

File: source/fe/fe_values.cc [hunk 1]
```diff
@@ -16,8 +16,10 @@
 #include <deal.II/base/array_view.h>
 #include <deal.II/base/memory_consumption.h>
 #include <deal.II/base/multithread_info.h>
+#include <deal.II/base/numbers.h>
 #include <deal.II/base/quadrature.h>
 #include <deal.II/base/signaling_nan.h>
+#include <deal.II/differentiation/ad.h>
```
Comment (jppelteret) [inline source/fe/fe_values.cc]: Note that if we use the unifying header then we avoid having to use this (potentially increasing) list of headers:
```
#include <deal.II/differentiation/ad/ad_number_traits.h>
#include <deal.II/differentiation/ad/adolc_number_types.h>
#include <deal.II/differentiation/ad/adolc_product_types.h>
#include <deal.II/differentiation/ad/sacado_number_types.h>
#include <deal.II/differentiation/ad/sacado_product_types.h>
```
It also means that if we add support for other numbers then we only need to maintain a this single header (plus whatever lists are in the build system) in order to ensure the compatibility of these classes.
  
Do you agree with my decision to use the unifying header, or would we prefer to keep the individual headers in place?

File: source/fe/fe_values.cc [hunk 2]
```diff
@@ -473,15 +473,19 @@ namespace FEValuesViews
       AssertDimension (values.size(), n_quadrature_points);
 
       std::fill (values.begin(), values.end(),
-                 dealii::internal::NumberType<typename std::decay<Number>::type>::value(0.0));
+                 dealii::internal::NumberType<Number>::value(0.0));
```
Comment (jppelteret) [inline source/fe/fe_values.cc]: @bangerth This changes attends to your concern that you raised in [this comment](https://github.com/dealii/dealii/pull/4951/files#r135063037). If this patch is accepted then I consider my obligation to answer your follow-up request, which I've sheepishly ignored, nullified ;-)

File: source/fe/fe_values.cc (hunk 1 above)
Comment (davydden) [reply source/fe/fe_values.cc]: I am fine to keep a single one which has all AD types in.

File: source/fe/fe_values.cc (hunk 2 above)
Comment (bangerth) [reply source/fe/fe_values.cc]: I appreciate it ;-)

**Comment by bangerth** [comment]
What an interesting observation!

The patch doesn't currently compile, but I'm ok with it in principle. This is such a surprising feature that I'd love to have a test that verifies this, if you think that one is easy enough to construct.

**Comment by jppelteret** [comment]
@bangerth I can't reproduce this build error, and I see that it also builds on the `tjhei-alpha`. This namespace is obviously been defined in `#include <differentiation/ad.h>`.  Do you have any suggestions regarding what the problem might be?

```
FAILED: /usr/bin/c++   -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_USE_DEBUG -I/home/bob/source/source/fe -Isource/fe -Iinclude -I/home/bob/source/include -I/home/bob/source/bundled/tbb41_20130401oss/include -I/home/bob/source/bundled/umfpack/UMFPACK/Include -I/home/bob/source/bundled/umfpack/AMD/Include -I/home/bob/source/bundled/boost-1.62.0/include -I/home/bob/source/bundled/muparser_v2_2_4/include -I/home/bob/libs/parmetis-4.0.3/include -I/home/bob/libs/oce-OCE-0.17/include/oce    -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Wno-literal-suffix -std=c++11 -Wno-parentheses -Wno-unused-local-typedefs -Og -ggdb -Wa,--compress-debug-sections -MMD -MT source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o -MF "source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o.d" -o source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o -c /home/bob/source/source/fe/fe_values.cc
/home/bob/source/source/fe/fe_values.cc: In function â€˜void dealii::FEValuesViews::internal::do_function_values(const dealii::ArrayView<Number>&, const dealii::Table<2, double>&, const std::vector<typename dealii::FEValuesViews::Scalar<dim, spacedim>::ShapeFunctionData>&, std::vector<typename dealii::ProductType<Number, double>::type>&)â€™:
/home/bob/source/source/fe/fe_values.cc:486:18: error: â€˜Differentiationâ€™ has not been declared
             if (!Differentiation::AD::is_ad_number<Number>::value)
```

**Comment by jppelteret** [comment]
@bangerth Making such a test would be doable - I think I've got an unrolled form of these functions somewhere in my commit history. I'll add it to those that will be upstreamed as tests for the main interface.

**Comment by bangerth** [comment]
It's not necessary that the test verifies that the result is correct, just that it's non-zero. Just create a test where the input vector is all zeros and test that the sensitivities are non-zero. That should be easy enough to do.

**Comment by bangerth** [comment]
As for the error: I'm a bit stumped. Can you try with GCC or do you only have clang on your system? The fact that clang succeeds is bizarre for this kind of error, but I don't know where to start poking. I just don't want to merge something that's almost certainly going to break for most everyone else, and on top of that break `tjhei-gcc-serial` for all following pull requests...

**Comment by jppelteret** [comment]
This works on both `clang 4.0.0` (or whatever Apple shipped with MacOS 10.11) and `gcc 6.3.0`. Ok, so at least I'm not (necessarily) missing something obvious. Since I can't reproduce this here, I'm going to try a couple of random things to see if I can it working on the tester.

File: tests/sacado/zero_val_nonzero_deriv_01.cc [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+// Tests the computation of the first derivatives of a function with zero
```
Comment (jppelteret) [inline tests/sacado/zero_val_nonzero_deriv_01.cc]: @bangerth Here's a test for you. Did I manage to capture the point that you were interested in?

Comment (bangerth) [reply tests/sacado/zero_val_nonzero_deriv_01.cc]: Yes, precisely!

**Comment by jppelteret** [comment]
@bangerth Oh, I know what's going on here. I've guarded the entire differentiation namespace with
```c++
#if defined(DEAL_II_WITH_ADOLC) || defined(DEAL_II_WITH_TRILINOS)
```
I think that it would be a good idea to define a macro such as
```c++
#defined DEAL_II_WITH_AD DEAL_II_WITH_ADOLC || DEAL_II_WITH_TRILINOS // ... and others in the future
```
which I could then easily use to guard these parts of the code.
If you agree that this is a good idea then where would you suggest I define this macro?

**Comment by jppelteret** [comment]
Actually that would still look horrible
```c++
#ifdef DEAL_II_WITH_AD
if (!Differentiation::AD::is_ad_number<Number>::value)
#endif
    if (value == dealii::internal::NumberType<Number>::value(0.0) )
        continue;
```
or
```c++
#ifdef DEAL_II_WITH_AD
if (!Differentiation::AD::is_ad_number<Number>::value)
    if (value == dealii::internal::NumberType<Number>::value(0.0) )
        continue;
#else
    if (value == dealii::internal::NumberType<Number>::value(0.0) )
        continue;
#endif
```
Thoughts / suggestions?

**Comment by bangerth** [comment]
`DEAL_II_WITH_AD_TYPES`?

**Comment by jppelteret** [comment]
Or else
```c++
namespace internal // in fe_values.cc
{
  namespace
  {
#ifdef DEAL_II_WITH_AD_TYPES
     template<typename Number>
     bool perform_value_check()
     {
      return !Differentiation::AD::is_ad_number<Number>::value;
    }
#else
     template<typename Number>
     bool perform_value_check()
     {
      return true;
    }
#endif
  }
}

// ... In a function
{
if (internal::perform_value_check<Number>())
    if (value == dealii::internal::NumberType<Number>::value(0.0) )
        continue;
```

I think that I like this approach the best

**Comment by bangerth** [comment]
Why not unconditionally declare the namespace and the `is_ad_number` member in it? If there are no AD types, it would simply always return false.

**Comment by jppelteret** [comment]
Yes, I suppose that would ultimately be the most robust solution.

**Comment by jppelteret** [comment]
Funny story... I do what you suggest in this already in `differentiation/ad/ad_number_[types/traits].h`, but I added in the strict guards to `differentiation/ad.h` last night without thinking too much about it :-/ Hopefully my next push should resolve this issue altogether.

**Comment by jppelteret** [comment]
Failing tests appear to be unrelated:
```
The following tests FAILED:
	316 - all-headers/matrix_free/fe_evaluation.h.debug (Timeout)
	317 - all-headers/matrix_free/mapping_info.templates.h.debug (Timeout)
Errors while running CTest
```

**Comment by bangerth** [comment]
OK, let's see what happens!

**Pull request merged**